design/assets.unity.md

```diff
@@ -0,0 +1,157 @@
+2. Maximize asset reuse, so engines can instance and optimize draw calls.
+
+
+IAssetCache
+------------
+
+This is a system-wide class instance that is responsible for caching assets beyond the lifetime of a single MRE
+instance. This could possibly be backed by persistent storage instead of memory. This is primarily intended
+for assets loaded via an HTTP request.
+
+### StoreAssets
+
+```cs
+void StoreAssets(
+	string uri,
+	IEnumerable<UnityEngine.Object> assets = null,
+	string version = null
+);
+```
+
+If all arguments are not null, and either the cache contains no assets for the resource, or the cached version
+does not match what is provided, this method stores all provided assets in the cache, overwriting any currently
+cached assets. Otherwise, the assets and version arguments are ignored, and the internal reference counter for this
+resource is decremented.
+
+* `uri` - The resource identifier
+* `assets` - The collection of assets generated from the given resource
+* `version` - The version of the loaded resource. Will typically be the HTTP response's `ETag` header.
+
+### LeaseAssets
+
+```cs
+Task<IEnumerable<UnityEngine.Object>> LeaseAssets(
+	string uri,
+	string ifMatches = null
+);
+```
+
+Asynchronously return the cached assets at the given URI, and increment the internal reference counter for this
+resource. Will return `null` if no assets are cached for that resource, or if `ifMatches` does not match the
+stored assets' `version`. This needs to be async in case the asset needs to be loaded from persistent storage.
+
+* `uri` - The resource identifier
+* `ifMatches` - If provided, assets will be returned only if the resource version matches this argument.
+
+### GetVersion
+
+```cs
+string GetVersion(string uri);
+```
+
+Returns the stored `version` of the given resource, or `null` if not cached. We'll need this for `If-Not-Match`
+HTTP headers.
+
+
+AssetManager
+---------------
+
+This class is responsible for maintaining references to active instances of assets used by a particular MRE instance.
+Every instance has its own AssetManager.
+
+### AssetMetadata
+
+```cs
+public struct AssetMetadata {
+	Guid Id;
+	Guid ContainerId;
+	UnityEngine.Object Asset;
+	ColliderGeometry ColliderGeometry;
+	AssetSource Source;
+}
+```
+
+Stores all the necessary info about an asset, including where it came from. `Source` will be null if this is not
+a shared asset, i.e. is a one-off creation of an MRE, or is a modified copy of something from the asset cache.
+
+### CacheRootGO
+
+The game object in the scene hierarchy that should be used as parent for any assets that require one, i.e. Prefabs.
+
+### EmptyTemplate
+
+The game object that should be duplicated for new actors.
+
+### GetById
+
+```cs
+AssetMetadata GetById(Guid? id, bool writeSafe = false);
+```
+
+Retrieve an asset by ID. If a write-safe asset is requested, and the stored asset with that ID is shared,
```

Comment:
Since the behavior is different depending on the parameter, would be better to have 2 methods instead.

Reply:
I think a single method makes sense. If an asset is being fetched as writeSafe, it only needs to be duplicated the first time. Subsequent calls will return the already-duplicated asset.